Add -Werror to compile GPDB

red1452 opened the pull request:
Add -Werror to compile GPDB
    
All warning were fixed.
1) checking for not-NULL was added before calling strlen().
2) converting long int to float8 was added for comparison between long int and
float8.
3) Unused variables were removed.
4) Useless memset's were removed from the insertsetbit function, because the buf
variable was zeroed before sending to the insertsetbit function.
5) Assert was fixed in the CreateDXLProjectNullsForGroupingSets function.
Execution of the Find method is extra, because it is checked below in the
condition and it may return nullptr.

andr-sokolov commented on configure.in:
```diff
@@ -2795,6 +2795,7 @@ fi
 if test "$GCC" = yes -a "$ICC" = no; then
   PGAC_PROG_CC_CFLAGS_OPT([-Werror=uninitialized])
   PGAC_PROG_CC_CFLAGS_OPT([-Werror=implicit-function-declaration])
+  PGAC_PROG_CC_CFLAGS_OPT([-Werror])
```
Can we delete previous 2 lines, because all warnings now turn into errors?

andr-sokolov replied:
-Werror=vla seems to be redundant too

andr-sokolov commented on contrib/dblink/dblink.c:
```diff
@@ -162,6 +162,8 @@ dblink_res_internalerror(PGconn *conn, PGresult *res, const char *p2)
 		PQclear(res);
 	elog(ERROR, "%s: %s", p2, msg);
+
+	exit(-1);
```
What was the warning here? My GCC 10.3.1 outputs no warning without `exit(-1)`

andr-sokolov commented on src/backend/access/transam/xact.c:
```diff
@@ -6939,7 +6939,8 @@ XactLogCommitRecord(TimestampTz commit_time,
 		XLogRegisterData((char *) (&xl_twophase), sizeof(xl_xact_twophase));
 		if (xl_xinfo.xinfo & XACT_XINFO_HAS_GID)
-			XLogRegisterData(unconstify(char *, twophase_gid), strlen(twophase_gid) + 1);
+			XLogRegisterData(unconstify(char *, twophase_gid),
```
Do we really need to call XLogRegisterData, when twophase_gid is NULL?

andr-sokolov commented on src/bin/pgbench/Makefile:
```diff
@@ -10,7 +10,7 @@ include $(top_builddir)/src/Makefile.global
 
 override CPPFLAGS := -I. -I$(srcdir) -I$(libpq_srcdir) $(CPPFLAGS)
-LDFLAGS_INTERNAL += -L$(top_builddir)/src/fe_utils -lpgfeutils $(libpq_pgport)
+LDFLAGS_INTERNAL += -L$(top_builddir)/src/fe_utils -lpgfeutils $(libpq_pgport) -lm
```
Is libm related to compiler warnings?

red1452 committed:
Add -Werror to compile GPDB

All warning were fixed.
1) checking for not-NULL was added before calling strlen().
2) converting LONG INT to FLOAT8 was added for comparison between LONG INT and
FLOAT8.

andr-sokolov commented:
typo in the description: LOAT8

andr-sokolov commented on configure.in:
```diff
@@ -2793,8 +2786,8 @@ fi
 if test "$GCC" = yes -a "$ICC" = no; then
-  PGAC_PROG_CC_CFLAGS_OPT([-Werror=uninitialized])
-  PGAC_PROG_CC_CFLAGS_OPT([-Werror=implicit-function-declaration])
+   PGAC_PROG_CC_CFLAGS_OPT([-Wall])
```
The -Wall flag has already been set here https://github.com/arenadata/gpdb/blob/15315c3cf96988e37868e9dbac1294f41ebfda7c/configure.in#L518

andr-sokolov commented on configure.in:
```diff
@@ -554,11 +552,6 @@ if test "$GCC" = yes -a "$ICC" = no; then
-  # some compilers (clang) don't recognize those and give spurious errors. Make
-  # sure the compiler supports fallthrough comments by explicitly requesting
-  # implicit-fallthrough level 3 (GCC's default).
-  PGAC_PROG_CC_CFLAGS_OPT([-Werror=implicit-fallthrough=3])
```
-Wimplicit-fallthrough=3 and -Wvla are not included into -Wall (https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html). Let's add these flags near -Wall

red1452 committed:
add -Wimplicit-fallthrough=3 -Wvla

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5786

andr-sokolov commented:
I get this error:
```
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-but-set-variable -Wno-format-truncation -Wno-stringop-truncation -O3  -Werror -I. -I. -I../../../../src/interfaces/libpq -I../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2   -c -o tsvector_op.o tsvector_op.c -MMD -MP -MF .deps/tsvector_op.Po
In function ‘tsCompareString’,
    inlined from ‘tsvector_bsearch’ at tsvector_op.c:1171:1:
tsvector_op.c:1188:9: error: ‘memcmp’ specified size between 18446744073709551612 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
 1188 |   cmp = memcmp(a, b, Min(lena, lenb));
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In function ‘tsCompareString’,
    inlined from ‘tsvector_bsearch’ at tsvector_op.c:1171:1,
    inlined from ‘tsvector_delete_arr’ at tsvector_op.c:625:13:
tsvector_op.c:1188:9: error: ‘memcmp’ specified size between 18446744073709551612 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
 1188 |   cmp = memcmp(a, b, Min(lena, lenb));
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Let's fix it too

red1452 commented:
> I get this error:
> 
> ```
> gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-but-set-variable -Wno-format-truncation -Wno-stringop-truncation -O3  -Werror -I. -I. -I../../../../src/interfaces/libpq -I../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2   -c -o tsvector_op.o tsvector_op.c -MMD -MP -MF .deps/tsvector_op.Po
> In function ‘tsCompareString’,
>     inlined from ‘tsvector_bsearch’ at tsvector_op.c:1171:1:
> tsvector_op.c:1188:9: error: ‘memcmp’ specified size between 18446744073709551612 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
>  1188 |   cmp = memcmp(a, b, Min(lena, lenb));
>       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> In function ‘tsCompareString’,
>     inlined from ‘tsvector_bsearch’ at tsvector_op.c:1171:1,
>     inlined from ‘tsvector_delete_arr’ at tsvector_op.c:625:13:
> tsvector_op.c:1188:9: error: ‘memcmp’ specified size between 18446744073709551612 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Werror=stringop-overflow=]
>  1188 |   cmp = memcmp(a, b, Min(lena, lenb));
>       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> Let's fix it too

It is bug at GCC less than 11: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91397
I do not think, that it is good idea to fix this. And I have not found the right way. You can add flag `-Wno-error=stringop-overflow` to your build.

andr-sokolov commented on src/backend/utils/adt/interpolate.c:
```diff
@@ -300,7 +300,7 @@ linterp_int64(PG_FUNCTION_ARGS)
 	{
 		r = round(y0+p*(y1-y0));
-		if ( r < LONG_MIN || r > LONG_MAX )
+		if ( r < (float8)LONG_MIN || r > (float8)LONG_MAX )
```
gcc 10 does not issue an error without this change

red1452 replied:
GCC in docker:
```
interpolate.c:303:28: warning: implicit conversion from 'long' to 'float8' (aka 'double') changes value from 9223372036854775807 to 9223372036854775808 [-Wimplicit-const-int-float-conversion]
  303 |                 if ( r < LONG_MIN || r > LONG_MAX )
      |                                        ~ ^~~~~~~~
/usr/bin/../lib/clang/17/include/limits.h:47:19: note: expanded from macro 'LONG_MAX'
   47 | #define LONG_MAX  __LONG_MAX__
      |                   ^~~~~~~~~~~~
<built-in>:56:22: note: expanded from macro '__LONG_MAX__'
   56 | #define __LONG_MAX__ 9223372036854775807L
      |                      ^~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

andr-sokolov commented on src/backend/gporca/libgpos/include/gpos/string/CWStringConst.h:
```diff
@@ -55,6 +55,10 @@ class CWStringConst : public CWStringBase
 
+	// There is the same Equals at the CWStringBase, add it here
+	// not to hidden it. [-Woverloaded-virtual]
+	using CWStringBase::Equals;
```
gcc 10 does not issue an error without this change

red1452 replied:
GCC 13 has:
```
g++ -std=c++14 -Werror -Wextra -Wpedantic -fno-omit-frame-pointer -Wall -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fexcess-precision=standard -O3 -I../../../../../../src/backend/gporca/libgpdbcost/include -I../../../../../../src/backend/gporca/libnaucrates/include -I../../../../../../src/backend/gporca/libgpopt/include -I../../../../../../src/backend/gporca/libgpos/include -I../../../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2   -c -o CLogger.o CLogger.cpp -MMD -MP -MF .deps/CLogger.Po
In file included from ../../../../../../src/backend/gporca/libgpos/include/gpos/string/CWString.h:14,
                 from ../../../../../../src/backend/gporca/libgpos/include/gpos/string/CWStringStatic.h:15,
                 from ../../../../../../src/backend/gporca/libgpos/include/gpos/error/CLogger.h:15,
                 from CLogger.cpp:12:
../../../../../../src/backend/gporca/libgpos/include/gpos/string/CWStringBase.h:79:22: error: ‘virtual gpos::BOOL gpos::CWStringBase::Equals(const gpos::WCHAR*) const’ was hidden [-Werror=overloaded-virtual=]
   79 |         virtual BOOL Equals(const WCHAR *str) const;
      |                      ^~~~~~
In file included from CLogger.cpp:20:
../../../../../../src/backend/gporca/libgpos/include/gpos/string/CWStringConst.h:63:14: note:   by ‘gpos::CWStringConst::Equals’
   63 |         BOOL Equals(const CWStringBase *str) const override;
      |              ^~~~~~
cc1plus: all warnings being treated as errors

```

andr-sokolov commented:
make[4]: вход в каталог «/home/keremet/compile/gpdb_arenadata/src/backend/access/bitmap»
gcc-13 -Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-but-set-variable -Wno-format-truncation -Wno-stringop-truncation -O3  -Werror -I../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2   -c -o bitmapinsert.o bitmapinsert.c -MMD -MP -MF .deps/bitmapinsert.Po
In file included from /usr/include/string.h:532,
                 from ../../../../src/include/c.h:63,
                 from ../../../../src/include/postgres.h:46,
                 from bitmapinsert.c:18:
In function ‘memset’,
    inlined from ‘insertsetbit’ at bitmapinsert.c:2148:2,
    inlined from ‘inserttuple.isra’ at bitmapinsert.c:2518:2:
/usr/include/bits/string_fortified.h:56:10: error: ‘__builtin___memset_chk’ forming offset [560, 1080] is out of the bounds [0, 560] of object ‘buf’ with type ‘BMTIDBuffer’ [-Werror=array-bounds=]
   56 |   return __builtin___memset_chk (__dest, __ch, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bitmapinsert.c: In function ‘inserttuple.isra’:
bitmapinsert.c:2447:25: note: ‘buf’ declared here
 2447 |         BMTIDBuffer     buf;
      |                         ^~~
cc1: all warnings being treated as errors

red1452 committed:
fix errors when optimization is o3

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5786

andr-sokolov commented:
CTranslatorQueryToDXL.cpp is compiled without -Werror

```
g++-13 -std=c++14 -Wall -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fexcess-precision=standard -O3 -I../../../../src/backend/gporca/libgpdbcost/include -I../../../../src/backend/gporca/libnaucrates/include -I../../../../src/backend/gporca/libgpopt/include -I../../../../src/backend/gporca/libgpos/include -I../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2   -c -o CTranslatorQueryToDXL.o CTranslatorQueryToDXL.cpp -MMD -MP -MF .deps/CTranslatorQueryToDXL.Po
In file included from ../../../../src/backend/gporca/libgpopt/include/gpopt/base/CColRef.h:16,
                 from ../../../../src/backend/gporca/libgpopt/include/gpopt/base/CUtils.h:16,
                 from CTranslatorQueryToDXL.cpp:33:
In member function ‘T* gpos::CHashMap<K, T, HashFn, EqFn, DestroyKFn, DestroyTFn>::Find(const K*) const [with K = int; T = unsigned int; gpos::ULONG (* HashFn)(const K*) = gpos::HashValue<int>; gpos::BOOL (* EqFn)(const K*, const K*) = gpos::Equals<int>; void (* DestroyKFn)(K*) = gpos::CleanupDelete<int>; void (* DestroyTFn)(T*) = gpos::CleanupDelete<unsigned int>]’,
    inlined from ‘gpdxl::CDXLNode* gpdxl::CTranslatorQueryToDXL::CreateDXLProjectNullsForGroupingSets(List*, gpdxl::CDXLNode*, gpos::CBitSet*, gpnaucrates::IntToUlongMap*, gpnaucrates::IntToUlongMap*, gpos::UlongToUlongMap*) const’ at CTranslatorQueryToDXL.cpp:4490:3:
../../../../src/backend/gporca/libgpos/include/gpos/common/CHashMap.h:240:44: warning: ‘this’ pointer is null [-Wnonnull]
  240 |                 CHashMapElem *elem = Lookup(key);
      |                                      ~~~~~~^~~~~
../../../../src/backend/gporca/libgpos/include/gpos/common/CHashMap.h: In member function ‘gpdxl::CDXLNode* gpdxl::CTranslatorQueryToDXL::CreateDXLProjectNullsForGroupingSets(List*, gpdxl::CDXLNode*, gpos::CBitSet*, gpnaucrates::IntToUlongMap*, gpnaucrates::IntToUlongMap*, gpos::UlongToUlongMap*) const’:
../../../../src/backend/gporca/libgpos/include/gpos/common/CHashMap.h:165:9: note: in a call to non-static member function ‘gpos::CHashMap<K, T, HashFn, EqFn, DestroyKFn, DestroyTFn>::CHashMapElem* gpos::CHashMap<K, T, HashFn, EqFn, DestroyKFn, DestroyTFn>::Lookup(const K*) const [with K = int; T = unsigned int; gpos::ULONG (* HashFn)(const K*) = gpos::HashValue<int>; gpos::BOOL (* EqFn)(const K*, const K*) = gpos::Equals<int>; void (* DestroyKFn)(K*) = gpos::CleanupDelete<int>; void (* DestroyTFn)(T*) = gpos::CleanupDelete<unsigned int>]’
  165 |         Lookup(const K *key) const
      |         ^~~~~~
```

red1452 committed:
add -Werror to CXXFLAGS

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5786

andr-sokolov commented on src/backend/access/bitmap/bitmapinsert.c:
```diff
@@ -2145,8 +2145,7 @@ insertsetbit(Relation rel, BlockNumber lovBlock, OffsetNumber lovOffset,
 	}
-	MemSet(buf->hwords, 0,
-		   BM_CALC_H_WORDS(buf->num_cwords) * sizeof(BM_HRL_WORD));
+	MemSet(buf->hwords, 0, sizeof(buf->hwords));
```
What about removing MemSets, because the function is called from one place, where buf has already been cleaned?

red1452 committed:
remove unused code

andr-sokolov commented on src/backend/cdb/motion/ic_udpifc.c:
```diff
@@ -1788,7 +1788,7 @@ destroyConnHashTable(ConnHashTable *ht)
 sendControlMessage(icpkthdr *pkt, int fd, struct sockaddr *addr, socklen_t peerLen)
 {
-	int			n;
+	int			n = 0;
```
This is redundant initialization. I suggest to delete this line and rewrite the loop at the end of the function:
```c
	/* retry 10 times for sending control message */
	for (int counter = 0; counter < 10; counter++)
	{
		int n = sendto(fd, (const char *) pkt, pkt->len, 0, addr, peerLen);
		if (n < 0)
		{
			if (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK)
				continue;

			write_log("sendcontrolmessage: got errno %d", errno);
			return;
		}

		if (n < pkt->len)
			write_log("sendcontrolmessage: got error %d errno %d seq %d", n, errno, pkt->seq);
		break;
	}
```

red1452 committed:
optimize code

andr-sokolov commented:
"memset was removed" -> "memset-s were removed"

red1452 committed:
small fix

red1452 committed:
format fix

RekGRpth commented on src/backend/access/transam/xact.c:
```diff
@@ -6938,7 +6938,7 @@ XactLogCommitRecord(TimestampTz commit_time,
 	{
 		XLogRegisterData((char *) (&xl_twophase), sizeof(xl_xact_twophase));
-		if (xl_xinfo.xinfo & XACT_XINFO_HAS_GID)
+		if (xl_xinfo.xinfo & XACT_XINFO_HAS_GID && twophase_gid)
```
XactLogAbortRecord function has same using strlen without checking

RekGRpth commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -4487,10 +4487,10 @@ CTranslatorQueryToDXL::CreateDXLProjectNullsForGroupingSets(
-			nullptr == sort_grouping_col_mapping,
-			nullptr != sort_grouping_col_mapping->Find(&sort_group_ref) &&
-				"Grouping column with no mapping");
+		GPOS_ASSERT(nullptr != sort_grouping_col_mapping &&
+					nullptr !=
+						sort_grouping_col_mapping->Find(&sort_group_ref) &&
+					"Grouping column with no mapping");
```
if below code calls AddSortingGroupingColumn function then sort_grouping_col_mapping->Find(&sort_group_ref) == nullptr and this assert raises before?

andr-sokolov replied:
```c
#define GPOS_ASSERT_IMP(x, y) GPOS_ASSERT(!(x) || (y))
```
so the equivalent expression in GPOS_ASSERT is
```
nullptr != sort_grouping_col_mapping || (nullptr != sort_grouping_col_mapping->Find(&sort_group_ref) &&
				"Grouping column with no mapping")
```

andr-sokolov commented on src/backend/utils/misc/guc.c:
```diff
@@ -5276,7 +5277,7 @@ find_option(const char *name, bool create_placeholders, int elevel)
 	 * the name field is first in config_generic.
 	 */
-	res = (struct config_generic **) bsearch((void *) &key,
+	res = (struct config_generic **) bsearch((void *) &ptr,
```
How can I reproduce the warning fixed here? GCC 10 and 13 don't find it

red1452 replied:
```
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-but-set-variable -Wno-format-truncation -Wno-stringop-truncation -O3  -Werror -I. -I. -I../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2  -DPG_KRB_SRVTAB='"FILE:/usr/local/gpdb/etc/postgresql/krb5.keytab"'  -c -o guc.o guc.c
In function ‘guc_var_compare’,
    inlined from ‘bsearch’ at /usr/include/x86_64-linux-gnu/bits/stdlib-bsearch.h:33:23,
    inlined from ‘find_option’ at guc.c:5279:35:
guc.c:5321:38: error: array subscript ‘const struct config_generic[0]’ is partly outside array bounds of ‘const char[8]’ [-Werror=array-bounds=]
 5321 |         return guc_name_compare(confa->name, confb->name);
      |                                 ~~~~~^~~~~~
guc.c: In function ‘find_option’:
guc.c:5267:25: note: object ‘name’ of size 8
 5267 | find_option(const char *name, bool create_placeholders, int elevel)
      |             ~~~~~~~~~~~~^~~~
cc1: all warnings being treated as errors
make[4]: *** [<builtin>: guc.o] Error 1
make[4]: Leaving directory '/home/evgeniy/gpdb/adb-7.x/src/backend/utils/misc'
make[3]: *** [../../../src/backend/common.mk:39: misc-recursive] Error 2
make[3]: Leaving directory '/home/evgeniy/gpdb/adb-7.x/src/backend/utils'
make[2]: *** [common.mk:39: utils-recursive] Error 2
make[2]: Leaving directory '/home/evgeniy/gpdb/adb-7.x/src/backend'
make[1]: *** [Makefile:45: all-backend-recurse] Error 2
make[1]: Leaving directory '/home/evgeniy/gpdb/adb-7.x/src'
make: *** [GNUmakefile:11: all-src-recurse] Error 2
evgeniy@evgeniy-pc:~/gpdb/adb-7.x$ gcc --version
gcc (Ubuntu 13.2.0-4ubuntu3) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

andr-sokolov commented:
"Initialization for variables were added." - where?

red1452 committed:
small fix

red1452 committed:
add brackets

red1452 committed:
run clang-format

andr-sokolov commented on src/backend/gpopt/translate/CTranslatorQueryToDXL.cpp:
```diff
@@ -4487,10 +4487,10 @@ CTranslatorQueryToDXL::CreateDXLProjectNullsForGroupingSets(
-				"Grouping column with no mapping");
+		GPOS_ASSERT(
+			nullptr != sort_grouping_col_mapping ||
+			(nullptr != sort_grouping_col_mapping->Find(&sort_group_ref) &&
```
The condition was rewritten correctly, but if sort_grouping_col_mapping is nullptr, then segfault happens

red1452 committed:
fix assert at CreateDXLProjectNullsForGroupingSets

bandetto commented on configure:
```diff
@@ -5418,7 +5418,7 @@ fi
 
 if test "$GCC" = yes -a "$ICC" = no; then
-  CFLAGS="-Wall -Wmissing-prototypes -Wpointer-arith"
+  CFLAGS="-Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla"
```
```suggestion
  CFLAGS="-Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough -Wvla"
```

Clang 18 says:
> warning: unknown warning option '-Wimplicit-fallthrough=3'; did you mean '-Wimplicit-fallthrough'? [-Wunknown-warning-option]

From https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html:

>    -Wimplicit-fallthrough is the same as -Wimplicit-fallthrough=3 and -Wno-implicit-fallthrough is the same as -Wimplicit-fallthrough=0.

Are there special meaning in using `-Wimplicit-fallthrough=3`?

bandetto reviewed (commented):
Build command was:
```
./configure CC=clang-18 CXX=clang++-18 --with-perl --with-python \
--enable-gpcloud --enable-orafce --enable-tap-tests --with-gssapi \
--with-openssl --with-perl --with-python --with-uuid=e2fs --with-llvm \
--without-gssapi --enable-depend --with-libxml --enable-debug \
--enable-debug-extensions --enable-cassert \
CFLAGS="-DEXTRA_DYNAMIC_MEMORY_DEBUG -DCDB_MOTION_DEBUG -g3 -fuse-ld=mold \
-fno-omit-frame-pointer -O0"
```

Warnings and errors from Clang 18:

## `-Wunused-but-set-variable`
1.
```
allpaths.c:2854:11: warning: variable 'pathkeys' set but not used [-Wunused-but-set-variable]
 2854 |         List       *pathkeys = NULL;
      |                     ^
```
2.
```
analyze.c:2293:15: warning: variable 'context' set but not used [-Wunused-but-set-variable]
 2293 |                 const char *context;
      |                             ^
```
3.
```
gram.y:3803:16: warning: variable 'fname' set but not used [-Wunused-but-set-variable]
 3803 |                                         Node               *fname;
      |
```
4.
```
gram.c:31422:9: warning: variable 'base_yynerrs' set but not used [-Wunused-but-set-variable]
 31422 |     int yynerrs = 0;
       |         ^
gram.c:72:25: note: expanded from macro 'yynerrs'
   72 | #define yynerrs         base_yynerrs
      |                         ^
```
5.
```
joinpath.c:2171:8: warning: variable 'have_nonmergeable_joinclause' set but not used [-Wunused-but-set-variable]
 2171 |         bool            have_nonmergeable_joinclause = false;
      |                         ^
```
6.
```
copy.c:2855:8: warning: variable 'num_phys_attrs' set but not used [-Wunused-but-set-variable]
 2855 |         int                     num_phys_attrs;
      |                                 ^
copy.c:2856:8: warning: variable 'attr_count' set but not used [-Wunused-but-set-variable]
 2856 |         int                     attr_count;
      |                                 ^
copy.c:5742:5: warning: variable 'attr_count' set but not used [-Wunused-but-set-variable]
 5742 |                                 attr_count;
      |                                 ^
```
7.
```
execUtils.c:1512:14: warning: variable 'currentSlice' set but not used [-Wunused-but-set-variable]
 1512 |         ExecSlice  *currentSlice;
      |                     ^
```
8.
```
explain.c:1490:14: warning: variable 'parentplanstate' set but not used [-Wunused-but-set-variable]
 1490 |         PlanState  *parentplanstate;
      |                     ^
```
9.
```
pgstat.c:4823:26: warning: variable 'xact_state' set but not used [-Wunused-but-set-variable]
 4823 |         PgStat_SubXactStatus   *xact_state;
      |                                 ^
```
10.
```
plancat.c:1631:24: warning: variable 'staForm' set but not used [-Wunused-but-set-variable]
 1631 |         Form_pg_statistic_ext staForm;
      |                               ^
```
11.
```
matview.c:174:10: warning: variable 'processed' set but not used [-Wunused-but-set-variable]
  174 |         uint64          processed = 0;
      |                         ^
```
12.
```
parse_utilcmd.c:2743:10: warning: variable 'found' set but not used [-Wunused-but-set-variable]
 2743 |                         bool            found = false;
      |                                         ^
```
13.
```
nodeHashjoin.c:1817:8: warning: variable 'nmoved' set but not used [-Wunused-but-set-variable]
 1817 |         int                     nmoved = 0;
      |                                 ^
```
14.
```
var.c:950:15: warning: variable 'attnum' set but not used [-Wunused-but-set-variable]
  950 |                         AttrNumber      attnum;
      |                                         ^
```
15.
```
nodeModifyTable.c:1769:11: warning: variable 'resultRelationDesc' set but not used [-Wunused-but-set-variable]
 1769 |         Relation        resultRelationDesc;
      |                         ^
```
16.
```
bitmaputil.c:1038:8: warning: variable 'init_page' set but not used [-Wunused-but-set-variable]
 1038 |         bool            init_page;
      |                         ^
```
17.
```
tablecmds.c:5978:9: warning: variable 'oldRelPersistence' set but not used [-Wunused-but-set-variable]
 5978 |                 char            oldRelPersistence;
      |                                 ^
tablecmds.c:5979:9: warning: variable 'oldAm' set but not used [-Wunused-but-set-variable]
 5979 |                 Oid                     oldAm;
      |                                         ^
tablecmds.c:17313:24: warning: variable 'conrec' set but not used [-Wunused-but-set-variable]
 17313 |                                 Form_pg_constraint conrec;
       |
```
18.
```
procarray.c:5282:9: warning: variable 'pid' set but not used [-Wunused-but-set-variable]
 5282 |         pid_t           pid;
      |                         ^
procarray.c:5283:12: warning: variable 'backendId' set but not used [-Wunused-but-set-variable]
 5283 |         BackendId       backendId;
      |                         ^
```
19.
```
extprotocolcmds.c:53:8: warning: variable 'protOid' set but not used [-Wunused-but-set-variable]
   53 |         Oid                     protOid;
      |                                 ^
```
20.
```
array_typanalyze.c:221:8: warning: variable 'null_cnt' set but not used [-Wunused-but-set-variable]
  221 |         int                     null_cnt = 0;
      |                                 ^
```
21.
```
nodeSubplan.c:1120:16: warning: variable 'explainRecvStats' set but not used [-Wunused-but-set-variable]
 1120 |         volatile bool explainRecvStats = false;
      |                       ^
```
22.
```
queue.c:590:11: warning: variable 'pVal' set but not used [-Wunused-but-set-variable]
  590 |                 Node                    *pVal;
      |                                          ^
queue.c:591:11: warning: variable 'ii' set but not used [-Wunused-but-set-variable]
  591 |                 int                              ii;
      |                                                  ^
```
23.
```
array_userfuncs.c:899:8: warning: variable 'lbs1' set but not used [-Wunused-but-set-variable]
  899 |                            *lbs1,
      |                             ^
array_userfuncs.c:903:8: warning: variable 'lbs2' set but not used [-Wunused-but-set-variable]
  903 |                            *lbs2,
      |                             ^
```
24.
```
nodeMotion.c:668:14: warning: variable 'outerPlan' set but not used [-Wunused-but-set-variable]
  668 |         PlanState  *outerPlan;
      |                     ^
```
25.
```
dfmgr.c:347:6: warning: variable 'lib_internal_version' set but not used [-Wunused-but-set-variable]
  347 |         int lib_internal_version = 0;
      |             ^
```
26.
```
xlog.c:2158:8: warning: variable 'npages' set but not used [-Wunused-but-set-variable]
 2158 |         int                     npages = 0;
      |                                 ^
```
27.
```
spi.c:2741:8: warning: variable 'eflags' set but not used [-Wunused-but-set-variable]
 2741 |         int                     eflags;
      |                                 ^
```
28.
```
nodeMotion.c:668:14: warning: variable 'outerPlan' set but not used [-Wunused-but-set-variable]
  668 |         PlanState  *outerPlan;
      |                     ^
```
29.
```
relcache.c:4403:8: warning: variable 'found' set but not used [-Wunused-but-set-variable]
 4403 |         int                     found;
      |                                 ^
```
30.
```
ic_tcp.c:1320:29: warning: variable 'pEntry' set but not used [-Wunused-but-set-variable]
 1320 |                 ChunkTransportStateEntry *pEntry = NULL;
      |                                           ^
```
31.
```
ic_udpifc.c:2975:8: warning: variable 'incoming_count' set but not used [-Wunused-but-set-variable]
 2975 |         int                     incoming_count = 0;
      |                                 ^
ic_udpifc.c:2976:8: warning: variable 'outgoing_count' set but not used [-Wunused-but-set-variable]
 2976 |         int                     outgoing_count = 0;
      |                                 ^
```
32.
```
gfile.c:1106:10: warning: variable 'openMode' set but not used [-Wunused-but-set-variable]
 1106 |                 mode_t openMode;
      |                        ^
```
33.
```
cdbmutate.c:546:8: warning: variable 'attno' set but not used [-Wunused-but-set-variable]
  546 |         int                     attno = 1;
      |                                 ^
```
34.
```
jsonpath_gram.c:1179:9: warning: variable 'jsonpath_yynerrs' set but not used [-Wunused-but-set-variable]
 1179 |     int yynerrs = 0;
      |         ^
jsonpath_gram.c:72:25: note: expanded from macro 'yynerrs'
   72 | #define yynerrs         jsonpath_yynerrs
      |                         ^
```
35.
```
appendonly_compaction.c:691:14: warning: variable 'relname' set but not used [-Wunused-but-set-variable]
  691 |         const char *relname;
      |                     ^
appendonly_compaction.c:793:14: warning: variable 'relname' set but not used [-Wunused-but-set-variable]
  793 |         const char *relname;
      |                     ^
```
36.
```
appendonlywriter.c:147:12: warning: variable 'pg_aoseg_dsc' set but not used [-Wunused-but-set-variable]
  147 |         TupleDesc       pg_aoseg_dsc;
      |                         ^
appendonlywriter.c:388:12: warning: variable 'pg_aoseg_dsc' set but not used [-Wunused-but-set-variable]
  388 |         TupleDesc       pg_aoseg_dsc;
      |                         ^
```
37.
```
resscheduler.c:230:12: warning: variable 'queuename' set but not used [-Wunused-but-set-variable]
  230 |                 char                            *queuename;
      |                                                  ^
1 warning generated.
resscheduler.c:761:9: warning: variable 'lockResult' set but not used [-Wunused-but-set-variable]
  761 |         int32           lockResult = 0;
      |                         ^
```
38.
```
regress_gp.c:631:20: warning: variable 'rsi' set but not used [-Wunused-but-set-variable]
  631 |         ReturnSetInfo      *rsi;
      |                             ^
regress_gp.c:1920:16: warning: variable 'topxid' set but not used [-Wunused-but-set-variable]
 1920 |         TransactionId topxid;
      |                       ^
```
39.
```
pg_dumpall.c:895:5: warning: variable 'i_rqoid' set but not used [-Wunused-but-set-variable]
  895 |                                 i_rqoid;
      |                                 ^
```
40.
```
gpfdist.c:3393:8: warning: variable 'buf' set but not used [-Wunused-but-set-variable]
 3393 |         char *buf;
      |               ^
```
41.
```
common.c:134:22: warning: variable 'binfo' set but not used [-Wunused-but-set-variable]
  134 |                 BinaryUpgradeInfo *binfo;
      |                                    ^
```
42.
```
exprparse.c:1081:9: warning: variable 'expr_yynerrs' set but not used [-Wunused-but-set-variable]
 1081 |     int yynerrs = 0;
      |         ^
exprparse.c:72:25: note: expanded from macro 'yynerrs'
   72 | #define yynerrs         expr_yynerrs
      |                         ^
```
43.
```
info.c:454:8: warning: variable 'relkind' set but not used [-Wunused-but-set-variable]
  454 |         char            relkind;
      |                         ^
```
44.
```
pg_dump.c:8899:8: warning: variable 'i_attencoding' set but not used [-Wunused-but-set-variable]
 8899 |         int                     i_attencoding;
      |                                 ^
pg_dump.c:20122:13: warning: variable 'separator' set but not used [-Wunused-but-set-variable]
 20122 |                         char       *separator;
       |                                     ^
```
45.
```
controldata.c:41:1 warning generated.
8: warning: variable 'got_tli' set but not used [-Wunused-but-set-variable]
   41 |         bool            got_tli = false;
      |                         ^
controldata.c:42:8: warning: variable 'got_log_id' set but not used [-Wunused-but-set-variable]
   42 |         bool            got_log_id = false;
      |                         ^
controldata.c:43:8: warning: variable 'got_log_seg' set but not used [-Wunused-but-set-variable]
   43 |         bool            got_log_seg = false;
      |                         ^
controldata.c:60:8: warning: variable 'got_toast' set but not used [-Wunused-but-set-variable]
   60 |         bool            got_toast = false;
      |                         ^
controldata.c:74:10: warning: variable 'tli' set but not used [-Wunused-but-set-variable]
   74 |         uint32          tli = 0;
      |                         ^
controldata.c:75:10: warning: variable 'logid' set but not used [-Wunused-but-set-variable]
   75 |         uint32          logid = 0;
      |                         ^
controldata.c:76:10: warning: variable 'segno' set but not used [-Wunused-but-set-variable]
   76 |         uint32          segno = 0;
      |                         ^
```
46.
```
indexscan.c:91:14: warning: variable 'blkno' set but not used [-Wunused-but-set-variable]
   91 |         BlockNumber blkno;
      |                     ^
```
47.
```
stream.c:437:13: warning: variable 'i' set but not used [-Wunused-but-set-variable]
  437 |         int i;
      |             ^
```

## `-Wunused-function`:
1.
```
datetime.c:3869:1: warning: unused function 'fast_encode_date' [-Wunused-function]
 3869 | fast_encode_date(struct pg_tm * tm, char *str, int *pos)
      | ^~~~~~~~~~~~~~~~
```

## Errors
1.
```
COptTasks.cpp:990:15: error: variable length arrays in C++ are a Clang extension [-Werror,-Wvla-cxx-extension]
  990 |                 char msgbuf[length];
      |                             ^~~~~~
COptTasks.cpp:990:15: note: read of non-const variable 'length' is not allowed in a constant expression
COptTasks.cpp:989:7: note: declared here
  989 |                 int length = NAMEDATALEN * rel_stats->Size() + 200;
      |                     ^
```
2.
```
CTranslatorUtils.cpp:390:16: error: variable length arrays in C++ are a Clang extension [-Werror,-Wvla-cxx-extension]
  390 |         OID arg_types[total_args];
      |                       ^~~~~~~~~~
CTranslatorUtils.cpp:390:16: note: initializer of 'total_args' is not a constant expression
CTranslatorUtils.cpp:388:14: note: declared here
  388 |         const ULONG total_args = num_args + num_return_args;
      |                     ^
CTranslatorUtils.cpp:391:17: error: variable length arrays in C++ are a Clang extension [-Werror,-Wvla-cxx-extension]
  391 |         char arg_modes[total_args];
      |                        ^~~~~~~~~~
CTranslatorUtils.cpp:391:17: note: initializer of 'total_args' is not a constant expression
CTranslatorUtils.cpp:388:14: note: declared here
  388 |         const ULONG total_args = num_args + num_return_args;
      |                     ^
```
3.
```
llvmjit_inline.cpp:811:15: error: 'startswith' is deprecated: Use starts_with instead [-Werror,-Wdeprecated-declarations]
  811 |         if (!modpath.startswith("$libdir/"))
      |                      ^~~~~~~~~~
      |                      starts_with
/usr/lib/llvm-18/include/llvm/ADT/StringRef.h:261:19: note: 'startswith' has been explicitly marked deprecated here
  261 |     [[nodiscard]] LLVM_DEPRECATED(
      |                   ^
/usr/lib/llvm-18/include/llvm/Support/Compiler.h:155:50: note: expanded from macro 'LLVM_DEPRECATED'
  155 | #define LLVM_DEPRECATED(MSG, FIX) __attribute__((deprecated(MSG, FIX)))
      |                                                  ^
```
4.
```
src/s3utils.cpp:108:5: error: 'MD5_Init' is deprecated [-Werror,-Wdeprecated-declarations]
  108 |     MD5_Init(&this->c);
      |     ^
/usr/include/openssl/md5.h:49:1: note: 'MD5_Init' has been explicitly marked deprecated here
   49 | OSSL_DEPRECATEDIN_3_0 int MD5_Init(MD5_CTX *c);
      | ^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
  182 | #   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
      |                                                 ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
   62 | #     define OSSL_DEPRECATED(since) __attribute__((deprecated))
      |                                                    ^
src/s3utils.cpp:112:5: error: 'MD5_Update' is deprecated [-Werror,-Wdeprecated-declarations]
  112 |     MD5_Update(&this->c, data, len);
      |     ^
/usr/include/openssl/md5.h:50:1: note: 'MD5_Update' has been explicitly marked deprecated here
   50 | OSSL_DEPRECATEDIN_3_0 int MD5_Update(MD5_CTX *c, const void *data, size_t len);
      | ^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
  182 | #   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
      |                                                 ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
   62 | #     define OSSL_DEPRECATED(since) __attribute__((deprecated))
      |                                                    ^
src/s3utils.cpp:117:5: error: 'MD5_Final' is deprecated [-Werror,-Wdeprecated-declarations]
  117 |     MD5_Final(this->md5, &c);
      |     ^
/usr/include/openssl/md5.h:51:1: note: 'MD5_Final' has been explicitly marked deprecated here
   51 | OSSL_DEPRECATEDIN_3_0 int MD5_Final(unsigned char *md, MD5_CTX *c);
      | ^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
  182 | #   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
      |                                                 ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
   62 | #     define OSSL_DEPRECATED(since) __attribute__((deprecated))
      |                                                    ^
src/s3utils.cpp:125:5: error: 'MD5_Init' is deprecated [-Werror,-Wdeprecated-declarations]
  125 |     MD5_Init(&this->c);
      |     ^
/usr/include/openssl/md5.h:49:1: note: 'MD5_Init' has been explicitly marked deprecated here
   49 | OSSL_DEPRECATEDIN_3_0 int MD5_Init(MD5_CTX *c);
      | ^
/usr/include/openssl/macros.h:182:49: note: expanded from macro 'OSSL_DEPRECATEDIN_3_0'
  182 | #   define OSSL_DEPRECATEDIN_3_0                OSSL_DEPRECATED(3.0)
      |                                                 ^
/usr/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
   62 | #     define OSSL_DEPRECATED(since) __attribute__((deprecated))
      |                                                    ^
```

bandetto commented on configure.in:
```diff
@@ -515,7 +515,7 @@ fi
 
 if test "$GCC" = yes -a "$ICC" = no; then
-  CFLAGS="-Wall -Wmissing-prototypes -Wpointer-arith"
+  CFLAGS="-Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla"
```
Maybe still ignore `-Wno-deprecated-declarations`?
```suggestion
  CFLAGS="-Wall -Wmissing-prototypes -Wpointer-arith -Wimplicit-fallthrough=3 -Wvla -Wno-deprecated-declarations"
```

andr-sokolov commented on src/backend/utils/misc/guc.c:
```diff
@@ -5266,7 +5266,8 @@ add_placeholder_variable(const char *name, int elevel)
 find_option(const char *name, bool create_placeholders, int elevel)
 {
-	const char **key = &name;
+	struct config_generic confa = {.name = name};
+	struct config_generic *ptr= &confa;
```
```suggestion
	struct config_generic *key = &(struct config_generic){.name = name};
```

red1452 committed:
fix flag -Wimplicit-fallthrough, other small fixes

red1452 committed:
minor fixes

dkovalev1 committed:
Fix argument types to avoid error: ‘memcmp’ specified bound

dkovalev1 commented:
```
/usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Xclang -no-opaque-pointers -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -Wno-deprecated-non-prototype -O2  -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_CONSTANT_MACROS -D_GNU_SOURCE -I/usr/lib/llvm-10/include  -I../../../../src/include   -D_GNU_SOURCE -I/usr/include/libxml2  -flto=thin -emit-llvm -c -o llvmjit_types.bc llvmjit_types.c
In file included from /usr/lib/llvm-10/include/llvm/Analysis/ModuleSummaryAnalysis.h:17,
                 from llvmjit_inline.cpp:51:
/usr/lib/llvm-10/include/llvm/IR/ModuleSummaryIndex.h: In constructor ‘llvm::modulesummaryindex::ModuleSummaryIndex(bool, bool)’:
/usr/lib/llvm-10/include/llvm/IR/ModuleSummaryIndex.h:995:73: error: member ‘llvm::modulesummaryindex::Alloc’ is used uninitialized [-Werror=uninitialized]
  995 |       : HaveGVs(HaveGVs), EnableSplitLTOUnit(EnableSplitLTOUnit), Saver(Alloc) {
      |                                                                         ^~~~~
cc1plus: all warnings being treated as errors
make[2]: * [<builtin>: llvmjit_inline.o] Error 1
make[2]: Leaving directory '/home/denis/arena/gpdb7/src/backend/jit/llvm'
make[1]: * [Makefile:45: all-backend/jit/llvm-recurse] Error 2
make[1]: Leaving directory '/home/denis/arena/gpdb7/src'
make: *** [GNUmakefile:11: all-src-recurse] Error 2
denis@deli:~/arena/gpdb7$ /usr/bin/clang --version
Ubuntu clang version 16.0.6 (++20231112100455+7cbf1a259152-1~exp1~20231112100542.106)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
denis@deli:~/arena/gpdb7$ gcc --version
gcc (Ubuntu 13.1.0-8ubuntu1~20.04.2) 13.1.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
It's ubuntu 20 with manually upgraded gcc and clang.

red1452 committed:
add ignore tsCompareString to abi compare test
